### PR TITLE
[BG-7938] Key generation and cosigning take encrypted sharded keys as input

### PR DESCRIPTION
**Pull request by dannybitgo**
- No more private keys displayed on the command line unless the user really wants it. Generate and Sign now can take in an encrypted sharded key file as input, and will prompt the user for the passwords before executing.

- In the process of adding this functionality I also reworked some of the command names, parameters, and internal functions.

- XLM master seed is still stored in an unencrypted .json file, but at least it doesn't get displayed in the terminal. 

- Also removed the need for user to download and install bitgo-cli (although the bitgo npm module is still a dependency)

- Also updated the README to align with the changes in commands. A sharded key file will be the default way to store the master private key.

### CODE REVIEW

File: app/admin.js [hunk 1]
```diff
@@ -334,15 +484,282 @@ const handleVerification = co(function *(args) {
...
+
+
+/**
+ * Generate a random BIP32 root key, from a random 256-bit
+ * seed. The seed is split using Shamir Secret Sharing Scheme
+ * (SSSS), such that any M of N of the shares can be recombined to
+ * produce the seed. The SSSS shares are encrypted with N separate
+ * passwords, intended to be provided at run-time by separate individuals.
+ */
+const handleInitShardedKey = function(args) {
```
Comment (dannybitgo) [inline app/admin.js]: This function was copied over from bitgo-cli, and it has been edited slightly.

File: app/utils.js [hunk 2]
```diff
@@ -169,3 +178,218 @@ exports.signXrpWithPrivateKey = function(txHex, privateKey, options) {
     id: exports.computeBinaryTransactionHash(serialized)
   };
 };
+
+
+exports.UserInput = function(args) {
```
Comment (dannybitgo) [inline app/utils.js]: All these UserInput functions were copied over from bitgo-cli

File: app/utils.js [hunk 3]
```diff
@@ -169,3 +178,218 @@ exports.signXrpWithPrivateKey = function(txHex, privateKey, options) {
...
+                delete self[name];
+                if (required) {
+                    return self.getIntVariable(name, question, required, min, max)();
+                }
+            });
+    };
+};
+
+
+
```
Comment (dannybitgo) [inline app/utils.js]: All the rest of these Utils were copied over from bitgo-cli

File: app/admin.js [hunk 4]
```diff
@@ -334,15 +484,282 @@ const handleVerification = co(function *(args) {
...
+
+
+/**
+ * Decrypt a key from the JSON file produced by initkey
+ * Note: by default, initkey only produces one key, so this will usually recover a list containing 1 key.
+ *
+ * This code has been refactored from a function that orignally decrypted a file containing multiple keys
+ *
+ */
+const decryptShardedKey = function(args, onFinish) {
```
Comment (dannybitgo) [inline app/admin.js]: This function used to be handleRecoverKeys - but I changed the name because it is now being used by multiple other functions. Any time we must decrypt a sharded key file, it must go through this function. So it is used by Sign, Generate, and Recover

File: app/admin.js [hunk 5]
```diff
@@ -7,7 +7,13 @@ const ArgumentParser = require('argparse').ArgumentParser;
 const pjson = require('../package.json');
 const fs = require('fs');
 const crypto = require('crypto');
-
+const Q = require('q');
```
Comment (dacruz21) [inline app/admin.js]: Q is pretty ancient, can you replace this with bluebird, or just vanilla promises if you can get away with it?

File: app/admin.js [hunk 6]
```diff
@@ -7,7 +7,13 @@ const ArgumentParser = require('argparse').ArgumentParser;
...
+const readline = require('readline');
+const read = require('read');
+const sjcl = require('sjcl');
+const secrets = require('secrets.js-grempe');
+const bitgo = require('bitgo');
+const bitcoin = bitgo.bitcoin;
```
Comment (dacruz21) [inline app/admin.js]: Let's use bitgo-utxo-lib instead.

File: app/admin.js [hunk 7]
```diff
@@ -101,17 +130,17 @@ setVerificationCommand.addArgument(
...
+  ['inputfile'],
   {
     action: 'store',
-    help: 'master private key to derive hardened child keys from'
+    help: 'file containing the master seed to derive hardened child keys from'
   }
 );
 generateKeysCommand.addArgument(
-  ['output'],
+  ['outputprefix'],
```
Comment (dacruz21) [inline app/admin.js]: I think this makes more sense as "outputfile", and if they specify a path that doesn't end in '.json', we can just append it for them.

File: app/admin.js [hunk 8]
```diff
@@ -133,22 +162,102 @@ generateKeysCommand.addArgument(
...
+initKey.addArgument(
+    ['--type'],
+    {
+        action: 'store',
+        defaultValue: 'shards',
+        help: 'pass in type "xlm" to create an xlm seed. note that xlm seeds are stored in a non-encrypted .json file'
+    }
+);
+
+const recoverKeys = subparsers.addParser('recover', {
```
Comment (dacruz21) [inline app/admin.js]: Do we need this?

File: app/admin.js [hunk 9]
```diff
@@ -253,8 +362,39 @@ const handleDeriveKey = function(args) {
...
 
 const handleGenerateKeys = function(args) {
+    if(args.type == 'xlm') {
+        handleGenerateXLMKeys(args);
+    } else {
+        handleGenerateFromShards(args);
+    }
+}
+
+const handleGenerateXLMKeys = function(args) {
```
Comment (dacruz21) [inline app/admin.js]: Up to you, we could encrypt the seed with SJCL.

File: app/admin.js [hunk 10]
```diff
@@ -334,15 +484,282 @@ const handleVerification = co(function *(args) {
...
+            return;
+        }
+        const passwordName = 'password' + i;
+        return input.getPassword(passwordName, 'Password for share ' + i + ': ', true)()
+            .then(function() {
+                return getPassword(i + 1, n);
+            });
+    };
+
+    return Q().then(function() {
```
Comment (dacruz21) [inline app/admin.js]: Let's use a bluebird coroutine here.

**Comment by tylerlevine** [comment]
Looks ok to me from at a high level. Don't have time right now do to a thorough review however.

Most issues I found are relatively minor:
* Some clean up to take care of extra whitespace, leftover/outdated comments and commented out code
* Modernization of promise handling code to use bluebird coroutines or native async await, depending on your runtime target
* Most of our other projects use 2 space indents

Larger issue is that a test is failing:
```
      ✓ should retrieve verification info on a key
Unhandled rejection Error: Unable to find wallet key: xpub6ARXqCvahM4dyWYDSPZMiii32yt3DTETyWCLDRZpQR4zpU9q6VmBKySA91hsLjofoUjdKdqPCcC54mbpJBmGNsNKM1szecH56p7Vk1byadR
    at /Users/tjl/src/key-recovery-service-v2/app/admin.js:286:11
    at next (native)
    at tryCatcher (/Users/tjl/src/key-recovery-service-v2/node_modules/bluebird/js/release/util.js:16:23)
    at PromiseSpawn._promiseFulfilled (/Users/tjl/src/key-recovery-service-v2/node_modules/bluebird/js/release/generators.js:97:49)
    at Promise._settlePromise (/Users/tjl/src/key-recovery-service-v2/node_modules/bluebird/js/release/promise.js:574:26)
    at Promise._settlePromise0 (/Users/tjl/src/key-recovery-service-v2/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/Users/tjl/src/key-recovery-service-v2/node_modules/bluebird/js/release/promise.js:693:18)
    at Async._drainQueue (/Users/tjl/src/key-recovery-service-v2/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/Users/tjl/src/key-recovery-service-v2/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/Users/tjl/src/key-recovery-service-v2/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:672:20)
    at tryOnImmediate (timers.js:645:5)
    at processImmediate [as _immediateCallback] (timers.js:617:5)


Key:                    xpub6AHA9hZDN11k2ijHMeS5QqHx2KP9aMBRhTDqANMnwVtdyw2TDYRmF8PjpvwUFcL1Et8Hj59S3gTSMcUQ5gAqTz3Wd8EsMTmF3DChhqPQBnU
Master Key:             undefined
User Email:             tester@bitgo.com
Verification Info:      verify user's identity by signed letter delivered by carrier pigeon
```

File: package.json [hunk 11]
```diff
@@ -17,23 +17,32 @@
...
+    "prompt-sync-history": "^1.0.1",
     "prova-lib": "^0.2.9",
     "q": "^1.5.1",
+    "read": "^1.0.7",
+    "readline": "^1.3.0",
     "ripple-binary-codec": "^0.1.13",
     "ripple-hashes": "^0.3.1",
     "ripple-keypairs": "^0.10.1",
     "ripple-lib": "^1.0.0-beta.5",
+    "secrets": "^0.1.0",
```
Comment (dacruz21) [inline package.json]: 11 vulnerabilities according to `npm audit` -- can you please version bump?

File: config.js [hunk 12]
```diff
@@ -20,7 +20,7 @@ module.exports = {
   "host": "0.0.0.0",
   "port": 6833,
   "adminemail": "davidcruz@bitgo.com",
-  "mongouri": "mongodb://localhost:27017/key-recovery-service",
+  "mongouri": process.env.KRSV2_DBURL,
```
Comment (dacruz21) [inline config.js]: Why this change? Is the Mongo URI sensitive?

File: app/utils.js (hunk 2 above)
Comment (dacruz21) [reply app/utils.js]: Is this still urgent, or can we spend some time modernizing? If possible let's use class syntax

File: app/decryptshards.js [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+                return getPassword(i, n, shares);
+            });
+    };
+
+    return Q().then(function() {
+        if (args.verifyonly) {
+            console.log('Verify Split Keys');
+        }
+    })
+    //.then(input.getVariable('file', 'Input file (JSON): '))
```
Comment (dacruz21) [inline app/decryptshards.js]: Cruft

File: app/decryptshards.js [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+    //.then(input.getVariable('keys', 'Comma-separated list of key indices to recover: '))
+    // key indices has been deprecated ^
+        .then(function() {
+            // Grab the list of keys from the file
+            // const json = fs.readFileSync(input.inputfile)
```
Comment (dacruz21) [inline app/decryptshards.js]: Cruft

File: config.js (hunk 12 above)
Comment (dannybitgo) [reply config.js]: yeah its sensitive - the way i'm using mongo now, the uri includes the username and password

File: app/decryptshards.js [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+    let n = null;
+    let xpub = null;
+    while(true) {
+        const inputfile = prompt("Enter the file path containing a key share: ");
+        try {
+            const file = fs.readFileSync(inputfile);
+            const json = JSON.parse(file);
+
+            // Enforce file contents are correct
+            if(!json.m) throw new Error('File does not contain the required field: m');
```
Comment (dacruz21) [inline app/decryptshards.js]: Can we clean this up a little? I think you could make this a loop -- something like

```
['m', 'n', /* etc */].forEach( requiredField => {
    if (!json[requiredField]) {
        throw new Error('file does not contain ' + requiredField)
    }
})
```

File: app/admin.js [hunk 16]
```diff
@@ -334,15 +430,234 @@ const handleVerification = co(function *(args) {
   }
 });
 
+const assertFileDoesNotExist = function(filename) {
+    try {
```
Comment (dacruz21) [inline app/admin.js]: there are `fs.exists()` and `fs.existsSync()` functions in node -- would make a lot more sense to use those

**Comment by marktoda** [comment]
@dannybitgo whats up with this? should we kill it? or re-review and rebase?

**Comment by dannybitgo** [comment]
kill
